### ISSUE TITLE
K3S provisioning support for arm64

### ISSUE THREAD

**Issue by thedadams** [rancher/rancher#34658]
**Is your feature request related to a problem? Please describe.**
The new K3S provisioning does not work for custom nodes with arm64 architecture. It seems that the `system-agent` supports arm64, but there is no version of `system-agent-installer` for arm64. Since both `system-agent` and k3s support arm64, it seems likely that adding a drone pipeline for the `system-agent-installer` would be enough to support arm64.

SURE-5295

**Comment by snasovich** [comment]
ARM64 is supported on RKE1 for Custom clusters, so we want this to be supported for RKE2 for feature parity.

**Comment by snasovich** [comment]
@Oats87 , after discussing it with @cloudnautique ARM64 support is out of scope for initial V2 cluster provisioning GA, so this is definitely lower priority.

**Comment by ubergeek801** [comment]
Hoping that some additional information might be helpful:

This is "almost" working for me (v2.6.0, ARM64 on both the Rancher server and the k3s node being created). The provisioner downloads the correct arch for `rancher-system-agent`, but ultimately the incorrect k3s binary is downloaded (`k3s` rather than `k3s-arm64`).

**Comment by kinarashah** [comment]
https://github.com/rancher/system-agent-installer-k3s/blob/main/.drone.yml

**Comment by thedadams** [comment]
Not stale.

**Comment by thedadams** [comment]
Still not stale.

**Comment by thedadams** [comment]
Not stale

**Comment by snasovich** [comment]
Added `kind/feature` label to tame stalebot here.

**Comment by sbstp** [comment]
After investigating how the various systems interact, I'm fairly confident that merging [this PR](https://github.com/rancher/system-agent-installer-k3s/pull/8) and triggering a new run of the pipeline would fix this issue. 

The `arm` and `arm64` images are already being created, tagged, and pushed to Docker Hub by the [system-agent-installer-k3s](https://github.com/rancher/system-agent-installer-k3s) pipelines. But they are not referenced in the docker manifest of the multi-arch image pushed to Docker Hub.

@HarrisonWAffel found the code in `rancher/rancher` which chooses the installer image to use and put in the `system-agent` plan. You can find this code [here](https://github.com/rancher/rancher/blob/57adc29c4f2c9dfa5b587494c00bec5bd22f68f3/pkg/provisioningv2/rke2/planner/planner.go#L969-L974).

This code resolves to image names such as `rancher/system-agent-installer-k3s:<version>"`. There is no additional logic to add the tag depending on the architecture, which means that the multi-arch image is being used.

But, as mentioned before, the multi-arch image has no references to `arm` or `arm64`  arches. So I'm pretty confident that merging that PR to references the arches in the multi-arch image will fix this issue.

**Comment by sbstp** [comment]
I was able to provision a cluster using the k3s arm64 image successfully using an arm64 ami & instance type on AWS. However there is still an issue with the `system-upgrade-controller`. The cluster will try to update and fail. I'm investigating this with Harrison's help.

The k3s folks have tagged new versions and the manifest was built correctly and references all the images as we wanted: https://hub.docker.com/r/rancher/system-agent-installer-k3s/tags

**Comment by Josh-Diamond** [comment]
Ticket #34658 - Test Plan
---

| Scenario | Test Case | Result |
|---|---|---|
| 1. | Deploy rancher server with local k3s and arm64 infrastructure ||
| 2. | Provision a downstream k3s Node driver cluster w/ arm64 infrastructure ||
| 3. | Provision a downstream k3s Custom cluster w/ arm64 infrastructure ||
| 4. | Import a downstream k3s cluster w/ arm64 infrastructure ||

**Comment by sbstp** [comment]
@HarrisonWAffel correctly identified that the issue was with the version of k3s I was using: it did not match the KDM data my rancher instance was using, and that sent the cluster into an update. By using an updated KDM ([from a branch](https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/dev-v2.7-1122-patches/data/data.json)) I was able to confirm that arm64 provisioning works well.

So at this point k3s 1.24.8 has been tagged and the arm/arm64 images are part of the multiarch manifest. Once KDM is updated to point to k3s 1.24.8, arm64 provisioning should work fine.

I don't think I can test non-64 bit arm easily. The best way I can think of testing it is to use an old Raspberry Pi, but I'm not sure I have the correct model for this.

**Comment by DaddyApi** [comment]
have the same issue. I have amd64 cluster, want to add arm workers,
fails
 level=error msg="error executing instruction 0: no child with platform linux/arm64 in index rancher/system-agent-installer-k3s:v1.24.4-k3s1: failed to get image index.docker.io/rancher/system-agent-installer-k3s:v1.24.4-k3s1"

**Comment by Josh-Diamond** [comment]
Ticket #34658 - Test Results - Re-opened
---

| Scenario | Test Case | Result |
|---|---|---|
| 1. | Deploy rancher server with local k3s and arm64 infrastructure | ❌ |
| 2. | Provision a downstream k3s Node driver cluster w/ arm64 infrastructure | ❌ |
| 3. | Provision a downstream k3s Custom cluster w/ arm64 infrastructure | ✅ |
| 4. | Import a downstream k3s cluster w/ arm64 infrastructure | ✅ |

---

With Docker on a single-node instance using rancher `v2.7-091ed163cc5c53efc50bd1a580cb4e54fa097e82-head`:

Scenario 1:
1. Fresh install of rancher `v2.7-head` using arm64 infrastructure - [`ami-0a9790c5a531163ee`/`ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220912`]
2. Verified - Rancher server successfully spins up and is accessible via UI
3. Veri-FAILED:
     - Rancher server is only able to create s3 compatible cloud credentials; all others fail w/ `a Config field must be set`, if attempted to be created through Node driver configurations
     - Linode Node driver is missing from Cluster Create options

Screenshots:

`S3 compatible credentials only`
<img width="1427" alt="Screenshot 2022-11-29 at 11 20 36 AM" src="https://user-images.githubusercontent.com/46494969/204613822-f3fa365b-874c-486a-a82c-9924380f0d8b.png">

`Failure to create cloud credentials from Node driver configs`
<img width="1201" alt="Screenshot 2022-11-29 at 11 21 09 AM" src="https://user-images.githubusercontent.com/46494969/204613951-7c9e628e-0c60-46d4-aeb9-50629643f4f1.png">

`Missing Linode Node driver`
<img width="1430" alt="Screenshot 2022-11-29 at 10 45 21 AM" src="https://user-images.githubusercontent.com/46494969/204613998-81addccc-d0e7-4236-b8e3-ed926d1d9000.png">

---

Scenario 2:
1. Provision a downstream Node driver k3s cluster w/ k8s `1.24.8+k3s1`
2. Veri-FAILED:
- Unable to create EC2 credentials
- Unable to use Linode
- As a result, unable to provision downstream Node driver clusters

---

Scenario 3:
1. Provision a downstream k3s Custom cluster w/ arm64 infrastructure - - [`ami-0a9790c5a531163ee`/`ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220912`]
2. Verified - Cluster successfully provisions + is functional

---

Scenario 4:
1. Import a downstream k3s cluster w/ arm64 infrastructure
2. Verified - Cluster successfully provisions + is functional

---

Additional Context:
Same failures were seen on rancher `v2.7.0`, however, on `v2.7.0` Linode Node driver was NOT missing from Cluster Create page.  Attempting to create cloud credentials for Linode on `v2.7.0` resulted in the same error seen on `v2.7-head`: `a Config field must be set`

**Comment by Josh-Diamond** [comment]
Failures above were related to Rancher (local) supporting k3s arm64, which is out-of-scope for this issue.

Re-testing using supported (local) architecture

**Comment by Josh-Diamond** [comment]
Ticket #34658 - Test Results [pt. 2]
---

| Scenario | Test Case | Result |
|---|---|---|
| 1. | Provision a downstream k3s Node driver cluster w/ arm64 infrastructure | ✅ |
| 2. | Provision a downstream k3s Custom cluster w/ arm64 infrastructure | ✅ |
| 3. | Import a downstream k3s cluster w/ arm64 infrastructure | ✅ |

---

With Docker on a single-node instance, using rancher `v2.7.0`:

Scenario 1:
1. Fresh install of rancher `v2.7.0`
2. Provision a downstream k3s Node driver cluster w/ k8s `v1.24.8+k3s1` on arm64 infrastructure 
    - AMI: [`ami-0a9790c5a531163ee`/`ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220912`]
    - Instance Type: [`a1.xlarge`]
3. Verified - Cluster successfully provisions and is functional

---

Scenario 2:
1. Provision a downstream k3s Custom cluster w/ arm64 infrastructure
2. Verified - Cluster successfully provisions and is functional

---

Scenario 3:
1. Import a downstream k3s cluster w/ arm64 infrastructure
2. Verified - Cluster successfully imported and functional

**Comment by sbstp** [comment]
Re-opening because the system agent update controller does not have an arm64 docker image.

PR to address: https://github.com/rancher/system-agent/pull/103

**Comment by snasovich** [comment]
@Josh-Diamond , the main functionality has been validated previously. I've created https://github.com/rancher/qa-tasks/issues/565 to track the additional change we've done after the validation was already done which can only be tested on Rancher version following the next release.
Please close the issue if you're OK with this.

cc: @sbstp @sowmyav27

**Comment by sowmyav27** [comment]
Waiting on this issue to move to-test to validate this - https://github.com/rancher/rancher/issues/40215

**Comment by Josh-Diamond** [comment]
Issue #40215 has been re-opened;

Moving this to `Blocked` until that issue has been resolved.

Re-open details:
https://github.com/rancher/rancher/issues/40215#issuecomment-1421036896

**Comment by Josh-Diamond** [comment]
Ticket #34658 - Test Results
---

Issue #40215 has been resolved - ✅ 

K3S provisioning support for arm64 has been validated, results [here](https://github.com/rancher/rancher/issues/40215#issuecomment-1426484515)
